### PR TITLE
JAMES-3406 James Distributed consistency model documentation page

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

**Review by nvduc91 — APPROVED** [review]

**Comment by chibenwa** [comment]
(Following discussions with Romaric!)

**Review by nvduc91 — APPROVED** [review]

**Review by LanKhuat — APPROVED** [review]

File: docs/modules/servers/pages/distributed/architecture/consistency-model.adoc [hunk 1]
```diff
@@ -1,19 +1,55 @@
...
+This page presents the consistency model used by the Distributed Server and
+points to the tools built around it.
 
 == Data Replication
 
-TODO
+The Distributed Server relies on different storage technologies, all having their own
+consistency models.
+
+These data stores replicates data in order to enforce some level of availability. We call
```
Comment (Arsnael) [inline docs/modules/servers/pages/distributed/architecture/consistency-model.adoc:12]:
```suggestion
These data stores replicate data in order to enforce some level of availability. We call
```

File: docs/modules/servers/pages/distributed/architecture/consistency-model.adoc [hunk 2]
```diff
@@ -1,19 +1,55 @@
...
+this process replication. By consistency, we mean the ability for all replica to hold the
+same data. By availability, we mean the ability for a replica to answer a request.
+
+In distributed systems, link:https://en.wikipedia.org/wiki/CAP_theorem[according to the CAP theorem],
+as we will necessarily encounter network partitions, then tradeoffs needs to be made between
```
Comment (Arsnael) [inline docs/modules/servers/pages/distributed/architecture/consistency-model.adoc:17]:
```suggestion
as we will necessarily encounter network partitions, then trade-offs need to be made between
```

File: docs/modules/servers/pages/distributed/architecture/consistency-model.adoc [hunk 3]
```diff
@@ -1,19 +1,55 @@
...
+consistency and availability.
+
+This section details this tradeoff for data stores used by the Distributed Server.
```
Comment (Arsnael) [inline docs/modules/servers/pages/distributed/architecture/consistency-model.adoc:20]:
```suggestion
This section details this trade-off for data stores used by the Distributed Server.
```

File: docs/modules/servers/pages/distributed/architecture/consistency-model.adoc [hunk 4]
```diff
@@ -1,19 +1,55 @@
...
 
 === Cassandra consistency model
 
-TODO
+link:https://cassandra.apache.org/[Cassandra] is an
+link:https://en.wikipedia.org/wiki/Eventual_consistency[eventually consistent] data store.
+This means that replica can hold diverging data, but are guaranteed to converge over time.
+
+Several mechanisms are built in Cassandra to enforce this convergence, and needs to be
```
Comment (Arsnael) [inline docs/modules/servers/pages/distributed/architecture/consistency-model.adoc:28]:
```suggestion
Several mechanisms are built in Cassandra to enforce this convergence, and need to be
```

File: docs/modules/servers/pages/distributed/architecture/consistency-model.adoc [hunk 5]
```diff
@@ -1,19 +1,55 @@
...
+leveraged by *Distributed Server Administrator*. Namely
+link:https://docs.datastax.com/en/dse/5.1/dse-admin/datastax_enterprise/tools/nodetool/toolsRepair.html[nodetool repair],
+link:https://cassandra.apache.org/doc/latest/operating/hints.html[Hinted hand-off] and
+link:https://cassandra.apache.org/doc/latest/operating/read_repair.html[Read repair].
+
+The Distributed Server tries to mitigate inconsistencies by relying on
+link:https://docs.datastax.com/en/archived/cassandra/3.0/cassandra/dml/dmlConfigConsistency.html[QUORUM] read and write levels.
+This means that a majority of replica are needed for read and write operations to be performed.
+
+Critical business operations, like UID allocation, relies on strong consistency mechanism brought by
```
Comment (Arsnael) [inline docs/modules/servers/pages/distributed/architecture/consistency-model.adoc:38]:
```suggestion
Critical business operations, like UID allocation, rely on strong consistency mechanisms brought by
```

File: docs/modules/servers/pages/distributed/architecture/consistency-model.adoc [hunk 6]
```diff
@@ -53,40 +53,66 @@ or very slow operations.
...
 
-=== Solve Inconsistency tasks
+While data can be consistent at the table level, some inconsistencies can sneak in at the applicative
+level across denormalization tables.
 
-TODO
+We write to a "table of truth" first, then duplicate the data to denormalization tables.
 
-=== Read repairs
+The Distributed server offers several mechanism to mitigate these inconsistencies:
```
Comment (Arsnael) [inline docs/modules/servers/pages/distributed/architecture/consistency-model.adoc:79]:
```suggestion
The Distributed server offers several mechanisms to mitigate these inconsistencies:
```

File: docs/modules/servers/pages/distributed/architecture/consistency-model.adoc [hunk 7]
```diff
@@ -53,40 +53,66 @@ or very slow operations.
...
+ - Cassandra is used for metadata storage
+ - ElasticSearch for search
+ - Object Storage for large object storage
+
+Thus the Distributed Server also offers mechanisms to enforce consistency across data stores.
 
 === Write path organisation
 
-TODO
+The primary data stores is composed of Cassandra for metadata and Object storage for binary data.
```
Comment (Arsnael) [inline docs/modules/servers/pages/distributed/architecture/consistency-model.adoc:100]:
```suggestion
The primary data stores are composed of Cassandra for metadata and Object storage for binary data.
```

File: docs/modules/servers/pages/distributed/architecture/consistency-model.adoc [hunk 8]
```diff
@@ -53,40 +53,66 @@ or very slow operations.
...
 
-=== ReIndexing
+This process, called the EventBus, which retries temporary errors, and stores transient errors for
+later admin-triggered retries is described further xref:distributed/operate/guide.adoc#_mailbox_event_bus[here].
+His role is to spread load and limit inconsistencies.
 
-TODO
\ No newline at end of file
+Furthermore, some xref:distributed/operate/guide.adoc#_usual_troubleshooting_procedures[re-indexing tasks]
+enables to re-synchronise ElasticSearch content with the primary data stores
```
Comment (Arsnael) [inline docs/modules/servers/pages/distributed/architecture/consistency-model.adoc:118]:
```suggestion
enable to re-synchronise ElasticSearch content with the primary data stores
```

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**